### PR TITLE
Rendering interpolator

### PR DESCRIPTION
**Pull request by Baccata**
Reworks the rendering logic to make use of a custom string-interpolator that allows to gather imports in an interpolated fashion. 

In addition, renames a few constructs with more intuitive semantics.

### CODE REVIEW

File: modules/codegen/src/smithy4s/codegen/ToLine.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      case Type.PrimitiveType(prim) => primitiveLine(prim)
+    }
+  }
+  private def primitiveLine(p: Primitive): Line =
+    p match {
+      case Primitive.Unit       => line"Unit"
+      case Primitive.ByteArray  => Line(Set("smithy4s.ByteArray"), "ByteArray")
+      case Primitive.Bool       => line"Boolean"
+      case Primitive.String     => line"String"
+      case Primitive.Timestamp  => line"smithy4s.Timestamp"
```
Comment (daddykotex) [inline modules/codegen/src/smithy4s/codegen/ToLine.scala:62]: don't we need the `Line(Set("smithy4s.Timestamp"), "Timestamp")` syntax here? and on like 71 too?

**Review by daddykotex — APPROVED** [review]
minor comments and question

File: modules/codegen/src/smithy4s/codegen/ToLine.scala (hunk 1 above)
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/ToLine.scala:62]: good catch

File: modules/codegen/src/smithy4s/codegen/PartialBlock.scala [hunk 2]
```diff
@@ -18,12 +18,15 @@ package smithy4s.codegen
...
 import cats.syntax.all._
 
-class PartialBlock(s: String) {
-  def apply[A](inner: A)(implicit A: Renderable[A]): RenderResult =
+class PartialBlock(l: Line) {
+  def apply[A](inner: A)(implicit A: ToLines[A]): Lines = {
+    val (imports, line) = l.tupled
     A.render(inner)
-      .transformLines(lines => (s + " {") :: indent(lines) ::: "}" :: Nil)
+      .transformLines(lines => (line + " {") :: indent(lines) ::: "}" :: Nil)
```
[the remark this answers is not in the document]
Comment (kubukoz) [reply modules/codegen/src/smithy4s/codegen/PartialBlock.scala:25]: I kinda prefer `++`, makes it more recognizable from `::` from the distance :grimacing:

I see more `:::` in the wild tho

File: modules/codegen/src/smithy4s/codegen/package.scala [hunk 3]
```diff
@@ -48,27 +59,35 @@ package object codegen {
...
+    new PartialBlock(Line(name))
   }
 
   private[codegen] def obj(
       name: String,
-      extensions: Seq[String]
+      extensions: List[Line]
   ): PartialBlock = {
-    obj(name, extensions.mkString(" with "))
+    obj(name, extensions.intercalate(Line(" with ")))
```
[the remark this answers is not in the document]
Comment (daddykotex) [reply modules/codegen/src/smithy4s/codegen/package.scala:90]: I checked out the code, and they exist in an external library if you ever care: https://github.com/scala/scala-collection-contrib

File: build.sbt [hunk 4]
```diff
@@ -295,8 +295,9 @@ lazy val codegen = projectMatrix
       Dependencies.Weaver.cats.value % Test
     ),
     testFrameworks += new TestFramework("weaver.framework.CatsEffect"),
-    scalacOptions := scalacOptions.value
+ scalacOptions := scalacOptions.value
```
Comment (Baccata) [inline build.sbt:298]:
```suggestion
    scalacOptions := scalacOptions.value
```

**Comment by Baccata** [comment]
good work @yisraelU !

**Pull request merged**